NitroxServer/GameLogic/Spawning/EntitySpawner.cs

```diff
@@ -1,118 +1,135 @@
 
-namespace NitroxServer.GameLogic
+namespace NitroxServer.GameLogic.Spawning
 {
     public class EntitySpawner
     {
-        private Dictionary<AbsoluteEntityCell, List<Entity>> entitiesByAbsoluteCell;
-
+        private readonly HashSet<Int3> parsedBatches = new HashSet<Int3>();
         private readonly Dictionary<string, WorldEntityInfo> worldEntitiesByClassId;
-        private readonly IEnumerable<EntitySpawnPoint> entitySpawnPoints;
         private readonly LootDistributionData lootDistributionData;
+        private readonly BatchCellsParser batchCellsParser;
+        private readonly Random random = new Random();
 
         private const uint TEXT_CLASS_ID = 0x31;
         private const uint MONOBEHAVIOUR_CLASS_ID = 0x72;
 
         public EntitySpawner()
         {
-            string lootDistributionString = "";
+            string lootDistributionString;
             if (GetDataFiles(out lootDistributionString, out worldEntitiesByClassId))
             {
-                BatchCellsParser BatchCellsParser = new BatchCellsParser();
-                entitySpawnPoints = BatchCellsParser.GetEntitySpawnPoints();
+                // TODO: If data files can't be loaded the code will crash due to NRE's.
+                batchCellsParser = new BatchCellsParser();
 
                 LootDistributionsParser lootDistributionsParser = new LootDistributionsParser();
                 lootDistributionData = lootDistributionsParser.GetLootDistributionData(lootDistributionString);
-
-                SpawnEntities();
             }
         }
 
-        public Dictionary<AbsoluteEntityCell, List<Entity>> GetEntitiesByAbsoluteCell()
+        /// <summary>
+        /// Spawns entities for <paramref name="batchId"/> if it was not loaded yet.
+        /// </summary>
+        public Optional<IDictionary<AbsoluteEntityCell, List<Entity>>> SpawnUnloadedEntitiesForBatch(Int3 batchId)
         {
-            return entitiesByAbsoluteCell;
-        }
+            lock (parsedBatches)
+            {
+                if (parsedBatches.Contains(batchId))
+                {
+                    return Optional<IDictionary<AbsoluteEntityCell, List<Entity>>>.Empty();
+                }
+                parsedBatches.Add(batchId);
+            }
 
-        private void SpawnEntities()
-        {
-            Log.Info("Spawning entities...");
-            entitiesByAbsoluteCell = new Dictionary<AbsoluteEntityCell, List<Entity>>();
-            Random random = new Random();
+            Log.Debug("Batch {0} not parsed yet; parsing...", batchId);
+
+            List<EntitySpawnPoint> batchSpawnPoints = batchCellsParser.ParseBatchData(batchId);
+
+            parsedBatches.Add(batchId);
+
+            Dictionary<AbsoluteEntityCell, List<Entity>> newSpawnPoints = new Dictionary<AbsoluteEntityCell, List<Entity>>();
 
-            foreach (EntitySpawnPoint entitySpawnPoint in entitySpawnPoints)
+            Parallel.ForEach(batchSpawnPoints, spawnPoint =>
```

Comment:
This method of multi threading has a lot of overhead associated with it in .Net. Have you bench-marked this to see if there is actually a performance benefit to concurrency/parallelism here? If we're locking on the same resources under the hood when we load game objects, then it may be faster to just serialize this process.

Reply:
You are right - I've blindly copied it over from the previous version without any consensus. A little profiling session indicates the deserialisation process (in `ParseBatchData`) takes the most CPU *by far*. The overhead of `Parallel.ForAll` and the `lock` in it fluctuates from 0 to 1 second on a non-mt test that takes 9 seconds (the test loads all batches in  `[8, 14)` for `x`, `y` and `z`). 

Currently batches are loaded in serial, when they are "touched" by some other code (as part of the lazy-loading process). To take optimal use of multithreaded processors, we may load all batches around it in parallel, high chance they'll be required soon anyway.
I don't think it's that relevant though (more code, more maintenance); when traversing the world very few new batches are needed every second, it seems fast enough to keep up for now.